adapters/huaweiads/huaweiads.go

```diff
@@ -0,0 +1,1395 @@
+	if err != nil {
+		return nil, err
+	}
+
+	bidder := &adapter{
+		endpoint:  config.Endpoint,
+		extraInfo: extraInfo,
+	}
+	return bidder, nil
+}
+
+func getExtraInfo(huaweiAdsExtraInfo string) (ExtraInfo, error) {
+	if huaweiAdsExtraInfo == "" {
+		return ExtraInfo{
+			TrackingUrl: DefaultTrackingEndpoint,
+		}, nil
+	}
+
+	var extraInfo ExtraInfo
+	if err := json.Unmarshal([]byte(huaweiAdsExtraInfo), &extraInfo); err != nil {
+		return extraInfo, fmt.Errorf("HuaweiAds ExtraInfo: invalid extra info: %v", err)
+	}
+
+	if extraInfo.TrackingUrl == "" {
+		extraInfo.TrackingUrl = DefaultTrackingEndpoint
+	}
+	return extraInfo, nil
+}
+
+// get request header
+func getHeaders(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds, request *openrtb2.BidRequest, isAddAuthorization bool) http.Header {
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	if huaweiAdsImpExt == nil {
+		return headers
+	}
+
+	if isAddAuthorization {
+		headers.Add("Authorization", getDigestAuthorization(huaweiAdsImpExt))
+	}
+
+	if request.Device != nil && len(request.Device.UA) > 0 {
+		headers.Add("User-Agent", request.Device.UA)
+	}
+	return headers
+}
+
+// get body json for HuaweiAds request
+func getHuaweiAdsReqJson(request *HuaweiAdsRequest, openRTBRequest *openrtb2.BidRequest, huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds) error {
+	request.Version = HuaweiAdxApiVersion
+	var err error
+	if err = getHuaweiAdsReqAppInfo(request, openRTBRequest); err != nil {
+		return err
+	}
+	if err = getHuaweiAdsReqDeviceInfo(request, openRTBRequest, huaweiAdsImpExt); err != nil {
+		return err
+	}
+	getHuaweiAdsReqNetWorkInfo(request, openRTBRequest)
+	getHuaweiAdsReqRegsInfo(request, openRTBRequest)
+	getHuaweiAdsReqGeoInfo(request, openRTBRequest)
+	return nil
+}
+
+// get adslot30
+func getHuaweiAdsReqAdslot30(huaweiAdsImpExt *openrtb_ext.ExtImpHuaweiAds,
+	openRTBImp *openrtb2.Imp, openRTBRequest *openrtb2.BidRequest) (Adslot30, error) {
+	adtypeLower := strings.ToLower(huaweiAdsImpExt.Adtype)
+	var adslot30 = Adslot30{
+		Slotid: huaweiAdsImpExt.SlotId,
+		Adtype: convertAdtypeString2Integer(adtypeLower),
+		Test:   int32(openRTBRequest.Test),
+	}
+
+	if openRTBImp.Banner != nil {
+		if openRTBImp.Banner.W != nil && openRTBImp.Banner.H != nil {
+			adslot30.W = *openRTBImp.Banner.W
+			adslot30.H = *openRTBImp.Banner.H
+		}
+		if len(openRTBImp.Banner.Format) != 0 {
+			var formats = make([]Format, 0, len(openRTBImp.Banner.Format))
+			for _, f := range openRTBImp.Banner.Format {
+				if f.H != 0 && f.W != 0 {
+					formats = append(formats, Format{f.W, f.H})
+				}
+			}
+			adslot30.Format = formats
+		}
+	} else if openRTBImp.Native != nil {
+		if err := getNativeFormat(&adslot30, openRTBImp); err != nil {
+			return adslot30, err
+		}
+	}
+
+	if adtypeLower == "roll" {
+		if openRTBImp.Video != nil {
+			adslot30.TotalDuration = int32(openRTBImp.Video.MaxDuration)
+		} else {
```

Comment:
Do you want to verify that `MaxDuration` is a positive number?

Reply:
In huaweiads document, roll ad need TotalDuration , but now only support banner native,  will support the roll ad later .

Reply:
corrected